src/pds/ingress/pds-nucleus-product-completion-checker.py

```diff
@@ -0,0 +1,321 @@
+    expected_data_files_table = dynamodb.Table('expected_data_files')
+
+    expected_data_files_table_get_response = expected_data_files_table.get_item(Key={
+            "s3_url_of_data_file": str(s3_url_of_data_file)
+        }
+    )
+
+    json_string_expected_files = json.dumps(expected_data_files_table_get_response, default=set_default)
+
+    expected_data_files_get_response_json = json.loads(json_string_expected_files)
+
+    expected_files_items = expected_data_files_get_response_json.get('Item')
+
+    return expected_files_items
+
+
+# Returns incomplete products
+def get_incomplete_products(s3_url_of_product_label):
+    incomplete_products_table = dynamodb.Table('incomplete_products')
+
+    incomplete_products_table_get_response = incomplete_products_table.get_item(Key={
+            "s3_url_of_product_label": str(s3_url_of_product_label)
+        }
+    )
+
+    json_string_incomplete_products = json.dumps(incomplete_products_table_get_response, default=set_default)
+    incomplete_products_table_get_response_json = json.loads(json_string_incomplete_products)
+    incomplete_products_table_items = incomplete_products_table_get_response_json.get('Item')
+    
+    return incomplete_products_table_items
+
+
+# Checks for s3_url_of_data_file in received files table
+def check_in_received_file_table(s3_url_of_data_file):
+    # Check received file table
+    received_data_files_table = dynamodb.Table('received_data_files')
+
+    received_data_files_table_get_response = received_data_files_table.get_item(Key={
+        "s3_url_of_data_file": str(s3_url_of_data_file)
+    })
+
+    json_string = json.dumps(received_data_files_table_get_response, default=set_default)
+    received_data_files_table_get_response_json = json.loads(json_string)
+    items = received_data_files_table_get_response_json.get('Item')
+
+    return items
+
+
+# Saves incomplete product label with a list if missing files
+def save_incomplete_product(s3_url_of_product_label, missing_files):
+    incomplete_products_table = dynamodb.Table('incomplete_products')
+
+    incomplete_products_table.put_item(
+        Item={
+            's3_url_of_product_label': s3_url_of_product_label,
+            'missing_files': missing_files
+        }
+    )
+
+
+# Updates exiting incomplete products
+def update_existing_incomplete_products(s3_url_of_data_file):
+    expected_files_items = get_expected_file_items(s3_url_of_data_file)
+
+    # If this is an already expected file (i.e.: another product was also expecting this)
+    if expected_files_items:
+        expected_files_product_label_list = expected_files_items['s3_urls_of_product_labels']
+        print("expected_files_product_label_list = " + str(expected_files_product_label_list))
+        
+        for expected_files_product_label in expected_files_product_label_list:
+
+            incomplete_products_table_items = get_incomplete_products(expected_files_product_label)
```

Comment:
I would have a function  `remove_expected_data_file_from_product(pruduct_key, data_file_key)` for the code inside the for loop.  which should lock the product record in the table `incomplete_product`. You would add a field 'locked' in this table and whenever this field is `True` the function `remove_expected_data_file_from_product` will start by waiting for the lock to be removed.

I guess you would use a dynamodb request like:

table.update_item (
  Key= ...
  updateExpression="SET locked = 1",
  ConditionExpession = "locked == 0"
)

And you would remove the lock when you finally update the missing_files in the record (as you do line 180).

Reply:
@tloubrieu-jpl , Thanks! I will handle locks in a separate pull  request.